### PR TITLE
TLS transport cache: don't cache transports for incomparable configs

### LINKED ISSUE #93409: Webhook REST clients cache invalid and possibly not expiring
**What happened**:
Deployed a webhook as a service to a Kubernetes cluster with 2 masters, listening initially on port 8000. 
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: webhook
---
kind: Service
apiVersion: v1
metadata:
  name: webhook
  namespace: webhook
spec:
  selector:
    app: webhook
  ports:
  - name: https
    protocol: TCP
    port: 8000
    targetPort: 443
```

Created a Validating Webhook pointing to port 8000:
```yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: validating-webhook
webhooks:
- clientConfig:
    caBundle: ...
    service:
      name: webhook
      namespace: webhook
      path: /validate
      port: 8000
  failurePolicy: Fail
...
```

Changed the service to use port 443. As expected, pod creation started to fail with the message (not always though):
```bash
$ kubectl run nginx --image nginx
Error from server (InternalError): Internal error occurred: failed calling webhook "webhook.webhook.svc": Post https://webhook.webhook.svc:443/validate?timeout=30s: no service port 'ὀ' found for service "webhook"
```

The 'ὀ' character was a bug on printing the port number - 8000 in this case - that was [fixed](https://github.com/kubernetes/kubernetes/pull/85558). This is not the issue here.

After fixing the webhook configuration, I started seeing the message:
```bash
$ kubectl run nginx --image nginx
Error from server (InternalError): Internal error occurred: failed calling webhook "webhook.webhook.svc": Post https://webhook.webhook.svc:443/validate?timeout=30s: no service port 'ƻ' found for service "webhook"
```

Consistently when hitting one of the masters, while the Validating Webhook worked as expected when hitting the other master. Again the weird character 'ƻ' represents 443, the .webhooks[].clientConfig.service.port in ValidatingWebhookConfiguration.

The same issue persists after deleting the service, webhook, and recreating with the same names.

**What you expected to happen**:
I expected the Validating Webhook to work on both masters.

**How to reproduce it (as minimally and precisely as possible)**:
Tried reproducing this issue using a different service name, as well as deploying it in another cluster. But the same error didn't happen again, so I believe there was some sort of special condition that caused this problem.

**Anything else we need to know?**:
Digging a bit into this, I suspect it has something to do with the [caching](https://github.com/kubernetes/kubernetes/blob/release-1.15/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L117) . I couldn't find a reference to where this cache is cleaned up.

Shouldn't we move to a LRUExpireCache to avoid this?

I'm happy to make a pull request if you think this is the case.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15 (but this code hasn't changed since)
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by roycaihw**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
tl;dr: Golang function values are not comparable. Using function addresses as part of the cache keys may cause the cache assigning stale transports to new clients with customzied configs (`Dial`, `Proxy` and `GetCert` in specific). The performance impact of restricting the cache is discussed in the special notes below. 

long story: When debugging #94810, I noticed that `TestValidateCachedClient` had [trouble](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/94704/pull-kubernetes-bazel-test/1304202091606052865) hitting the webhook server, e.g.

```
dial tcp 127.0.0.1:33847: connect: connection refused
```

Printing out the address shows the webhook controller was trying to hit the webhook server in the previous unittest `TestValidate`.

It turns out that even though
- the test [customized the webhook’s ServiceResolver](https://github.com/kubernetes/kubernetes/blob/v1.19.0/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/plugin_test.go#L182)
- which [customized the dialer](https://github.com/kubernetes/kubernetes/blob/v1.19.0/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L185-L187) in the rest client config

the webhook client still used the stale TLS transport from [TLSTransportCache](https://github.com/kubernetes/kubernetes/blob/v1.19.0/staging/src/k8s.io/client-go/transport/cache.go#L41), because
- the [function pointer](https://github.com/kubernetes/kubernetes/blob/v1.19.0/staging/src/k8s.io/client-go/transport/cache.go#L137) is used as part of the cache key
- and [function pointers denote the code of the function](https://stackoverflow.com/a/36127142), which didn’t change in the webhook case (the variables make the function dynamic, but the code is in the same place in memory)

Disabling the cache fixed the issue. Looking at the [tlsCacheKey](https://github.com/kubernetes/kubernetes/blob/v1.19.0/staging/src/k8s.io/client-go/transport/cache.go#L131-L140), the `GetCert`, `Dial` and `Proxy` functions have the same issue. 

**Which issue(s) this PR fixes**:
Fixes #93409, #94810

**Special notes for your reviewer**:
We will cherrypick this change to supported minor versions. 

Talked with @liggitt, a major concern is that restricting the cache may expose us to ephemeral port exhaustion (xref https://github.com/kubernetes/kubernetes/pull/11820). It turns out that [go1.7](https://golang.org/doc/go1.7#net_http) supports `IdleConnTimeout` which helps manage idle connection cleanup (it wasn't supported when the cache was implemented), and k8s has been [using](https://github.com/kubernetes/kubernetes/blob/05a46dbb6026dfd9ae11600f8dc08a99994acdec/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L121-L123) this feature. The cache still has value: e.g. in case of clientsets unnecessarily creating unique transports for different api groups. We added log to the code measuring transport creation in tests, to evaluate impact and future improvements. 

**Does this PR introduce a user-facing change?**:

```release-note
Fixed a bug in client-go where new clients with customized `Dial`, `Proxy`, `GetCert` config may get stale HTTP transports.
```

/sig api-machinery
/cc @liggitt

### CODE REVIEW

**Comment by liggitt** [comment]
downloading all the component logs from the e2e-gce-100-performance and e2e runs, and looking for the added message:

scale:
```
grep "created an uncacheable transport for UserAgent" * | grep -o "caller.* " | sort | uniq -c
  84 caller: cmd/kubelet/app/server.go:570 
   2 caller: cmd/kubelet/app/server.go:579 
  84 caller: cmd/kubelet/app/server.go:594 
  18 caller: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:281 
2109 caller: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:188 
```

e2e:
```
  84 caller: cmd/kubelet/app/server.go:570 
   2 caller: cmd/kubelet/app/server.go:579 
  84 caller: cmd/kubelet/app/server.go:594 
  45 caller: staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:151 
  18 caller: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:281 
1803 caller: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:188
```

The ones inside cmd are not concerning, since those happen once at start up.

handler_proxy.go:281 happens on update of APIService, which is slow-moving

util/webhook/client.go:151 caches based on the webhook config, which is slow-moving

available_controller.go:188 is the only questionable one I see. It looks like it fires every 30 seconds on every remote APIService and makes ~5 calls with that client
* most clusters range from 1-O(dozens) of remote APIService objects
* an idle connection timeout of 90 seconds would accumulate ~3 sync iterations of idle connections
* ~50 remote APIService objects * ~5 requests * ~3 sync iterations == ~750 idle connections because of this change

**Comment by liggitt** [comment]
cc @deads2k

**Comment by liggitt** [comment]
It looks like the transport constructed by the available_controller is not APIService specific, but can only vary on the dynamic cert/key. We could do something specific inside that controller to reuse the same restTransport as long as the result of `c.proxyCurrentCertKeyContent()` is the same

**Comment by roycaihw** [comment]
Added a local cache to available_controller

[e2e-gce-100-performance](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/95427/pull-kubernetes-e2e-gce-100-performance/1314715089328148480/artifacts/e2e-95427-95a39-master/):
```
$ grep "created an uncacheable transport for UserAgent" * | grep -o "caller.* " | sort | uniq -c
  84 caller: cmd/kubelet/app/server.go:570
   2 caller: cmd/kubelet/app/server.go:579
  84 caller: cmd/kubelet/app/server.go:594
  16 caller: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:281
   1 caller: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:111
```

[e2e-gce-ubuntu-containerd](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/95427/pull-kubernetes-e2e-gce-ubuntu-containerd/1314715089198125056/artifacts/e2e-9923e4960e-a7d53-master/):
```
  42 caller: cmd/kubelet/app/server.go:570
   1 caller: cmd/kubelet/app/server.go:579
  42 caller: cmd/kubelet/app/server.go:594
  36 caller: staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:151
  15 caller: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:281
   1 caller: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:111
```

**Comment by roycaihw** [comment]
> it fires every 30 seconds on every remote APIService

Just want to clarify. [30s](https://github.com/kubernetes/kubernetes/blob/1dff024e6449a98b6558c242b09e10f98efdb5ce/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L115) is the case when an APIService is [constantly failing discovery](https://github.com/kubernetes/kubernetes/blob/1dff024e6449a98b6558c242b09e10f98efdb5ce/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L438). If the discovery [succeeded](https://github.com/kubernetes/kubernetes/blob/1dff024e6449a98b6558c242b09e10f98efdb5ce/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L429-L435), the controller fires requests every [5 minutes](https://github.com/kubernetes/kubernetes/blob/1dff024e6449a98b6558c242b09e10f98efdb5ce/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L176).

Still, without caching it can be an issue. The controller has a fairly tight requeue time (5ms initially), which can cause many new connections been created in a short time.

**Comment by liggitt** [comment]
lgtm, drop the do-not-merge commit and squash the fixup

**Comment by roycaihw** [comment]
Squashed

/retest
